### ISSUE TITLE
newrelic_browser_application to return NREUM information

### ISSUE THREAD

**Issue by randompixel** [newrelic/terraform-provider-newrelic#2388]
https://registry.terraform.io/providers/newrelic/newrelic/latest/docs/resources/browser_application 

I have used this to create a browser application. I want to be able to automatically take the output of this and feed the variables from the Copy/Paste Javascript code box from Terraform and place them directly into https://github.com/newrelic/newrelic-browser-agent the build as part of the build pipeline

https://github.com/newrelic/newrelic-browser-agent says 
> From the Copy/Paste JavaScript box, copy the configuration values assigned to the NREUM object (init, info, and loader_config). You will use these configuration values when instantiating the agent using the NPM package.

Therefore would it be possible to export the values from the NREUM object as Terraform output which could then be consumed elsewhere? That way I can go from NR creation to application deployment without any manual interventions.

Hope this is possible, thanks in advance!

**Comment by pranav-new-relic** [comment]
Hi @randompixel - thank you for reporting this issue. Upon preliminary lookup, it does look like this is doable, but I'm not sure if that can exactly be in the format you've requested, i.e. the `init`, `info` and `loader_config` extracted from the JavaScript code snippet that can be exported by the Terraform resource.

For starters, the API that the Terraform Resource depends on seems to export an attribute `jsLoaderScript` in `browserProperties` that contains the entire code snippet seen in the box in the UI. I'd just want another pair of eyes to confirm if this is the same script that is seen in the code snippet - if you've used NerdGraph anytime, would you mind trying the following mutation out and letting me know if `jsLoaderScript` is what is seen in the code box? 

```gql
{
  actor {
    entity(guid: "") {
      accountId
      alertSeverity
      domain
      name
      ... on BrowserApplicationEntity {
        name
        accountId
        browserProperties {
          jsLoaderScript
        }
        guid
        applicationId
      }
      entityType
      guid
    }
  }
}
```

I find `init`, `info` and `loader_config` in the `NREUM` object in this attribute - but would like you to have a look too, if this is right :)

If the above is true - I'd be able to check with my team on when we can get this done, as it'll take quite a while to get this done :) thank you!

**Comment by pranav-new-relic** [comment]
(edited comment: changes made based on further lookup)

Hi @randompixel - thank you for reporting this issue. Upon preliminary lookup, it does look like this is doable, but I'd have to check with my team on this to confirm the feasibility of exporting these attributes from the Terraform resource, after it is created.

For starters, the API that the Terraform Resource depends on seems to export an attribute `jsConfig` in `browserProperties` that contains the attributes `init`, `info` and `loader_config`, and the contents of these attributes match the `init`, `info` and `loader_config` attributes seen in the JavaScript code snippet shown in the UI, in the the browser application. 

If you have used NerdGraph anytime, can you please try this mutation out (by supplying the GUID of your browser application) and let me know if the `jsConfig` attribute in the response comprises the aforementioned attributes and would hence, suffice this requirement?

```gql
{
  actor {
    entity(guid: "") {
      accountId
      alertSeverity
      domain
      name
      ... on BrowserApplicationEntity {
        name
        accountId
        browserProperties {
          jsConfig
        }
        guid
        applicationId
      }
      entityType
      guid
    }
  }
}

```
If the above is true - I'd be able to check with my team on when we can get this done, as it'll take quite a while to get this done :) thank you!

**Comment by pranav-new-relic** [comment]
Hi @randompixel - an update. It looks like this would not be feasible to implement soon, owing to a couple of more observations (one of which I missed in the last update - apologies about this).

The query I've posted above is used by the relevant Terraform resource to **read** an entity after it is created, in order to compare remote changes against local changes (in the state file), which is used to update an existing resource. However, the following mutation is used to **create** a new resource, which currently, does not export the three attributes we currently intend to export (as discussed above). As you may see in the image below, in this mutation, we do not have the `browserProperties` attribute which would comprise `jsConfig` with `init`, `info` and `loader_config` (unlike in the query discussed earlier in this thread).

<img width="285" alt="image" src="https://github.com/newrelic/terraform-provider-newrelic/assets/127438038/cc418b39-96e9-43c3-81fe-b68bdb52843b">

Given there is a dependency on the team building the relevant API for this (as this would have to be conveyed to them, who would then add it if these are found resourceful), after which we would be able to add this to Terraform, this would take a long while. Also, since the `info` attribute includes a `licenseKey`, it is still not clear if it would be right to export such (sensitive) attributes via the Terraform Resource.

To counter these, I've a workaround to suggest you to implement (that I've tried on my device and works fine, which would suffice your requirement). I've added details in the next comment.

**Comment by pranav-new-relic** [comment]
Since the idea is to retrieve attributes of the NREUM object (`init`, `info` and `loader_config`) of the Browser Application created via Terraform, the procedure to obtain these can be automated using the `graphql` provider. The idea is to create a browser application via Terraform using the NR Terraform Provider, use the GUID exported by this resource as the input to a GraphQL query that is sent via the `graphql_query` resource, that would help fetch required attributes, which are then beautified as needed.

You can try using the configuration below (to understand what I've mentioned above). All of these files would be in one folder, in which `terraform plan` and `terraform apply` can be run.

`query.gql`
```gql
query(
	$guid: EntityGuid!,
) { actor { entity(
	guid: $guid,
) {
	accountId
	domain
	guid
	name
	type
	... on BrowserApplicationEntity {
		__typename
		applicationId
		browserProperties {
			jsConfig
		}
	}
} } }
```

`variables.tf`
```hcl
variable "graphiql_url" {
  type = string
  default = "https://api.eu.newrelic.com/graphql"
}

variable "api_key" {
  type = string
  default = "XXXX-XXXXXXXXX (Your New Relic API Key)"
}
```

`main.tf`
```hcl
terraform {
  required_providers {
    newrelic = {
      source = "newrelic/newrelic"
    }
    graphql = {
      source = "sullivtr/graphql"
    }
  }
}

locals {
  response = jsondecode(data.graphql_query.sample_query.query_response)
  info = local.response["data"]["actor"]["entity"]["browserProperties"]["jsConfig"]["info"]
  init = local.response["data"]["actor"]["entity"]["browserProperties"]["jsConfig"]["init"]
  loader_config = local.response["data"]["actor"]["entity"]["browserProperties"]["jsConfig"]["loader_config"]
}

provider "newrelic" {
  region = "US" # US or EU
}

provider "graphql" {
  url = var.graphiql_url
  headers = {
    "Content-Type" = "application/json"
    "API-Key" = var.api_key
  }
}

resource "newrelic_browser_application" "foo" {
  name = "Sample Browser Application 2388"
  cookies_enabled = true
  distributed_tracing_enabled = true
  loader_type = "SPA"
}

data "graphql_query" "sample_query" {
  query_variables = {
    "guid" = newrelic_browser_application.foo.id
  }
  query  = file("query.gql")
  depends_on = [newrelic_browser_application.foo]
}

#uncomment these to see the retrieved attributes in the terminal
#output "required_attributes" {
#  value = {
#    "info": local.info,
#    "init": local.init,
#    "loader_config": local.loader_config
#  }
#}
#
#output "browser_application_guid" {
#  value = newrelic_browser_application.foo.id
#}
```

**Comment by randompixel** [comment]
OK, thanks @pranav-new-relic 

I'll try and wrap that up into a private Terraform module I can import into all my projects. It may take me a few days.

If it is too hard to implement the native functionality maybe the Terraform documentation could be updated to include a reference to solving this problem?

**Comment by pranav-new-relic** [comment]
Thanks for taking a look, @randompixel. Not that it is hard to implement - but it is tedious given this change would span across teams, and would need discussion, given the specificity of this use case, as opposed to which the workaround we've discussed should work (also because  it would probably not be ideal to export an attribute comprising a "license key").

I think the suggestion to include this in the docs is really useful, as such an addition would help users to leverage GraphQL within Terraform (that is not specific to just this use case, and would apply to any GraphQL query). I shall check with the team on this - thank you for the suggestion.

**Comment by pranav-new-relic** [comment]
Update: We've identified a solution to avoid the deficit of these fields we currently have in the API mutation, which can help get this implemented in the Terraform resource. We've added this to our task list, and will pick it up for development based on priority - in the meanwhile, you can make use of the workaround we've already discussed above.

**Comment by randompixel** [comment]
Hi @pranav-new-relic ,

That's excellent news. I will still have a go at your GraphQL solution in the interim.

Thanks for the update.

**Comment by pranav-new-relic** [comment]
Hi @randompixel, the solution to this has been released with v3.27.3 of the Terraform Provider. We shall add a note to the documentation on the usage of the attribute `js_config` using which browser properties may be exported by this resource; however, in the interim, kindly refer to the bottom of the description of [this PR](https://github.com/newrelic/terraform-provider-newrelic/pull/2421) for a sample of how these exported attributes may be used.

Thank you for reporting this issue :)
